connectors/sources/sharepoint_online.py

```diff
@@ -879,258 +1003,240 @@ async def validate_config(self):
+                            ]
+                        }
+                    }
+                }
+            }
+        """
 
-        list_item_role_assignments = await self.client.site_list_item_role_assignments(
-            site_web_url, site_list_name, list_item["id"]
-        )
+        if "UserName" in user:
+            username_field = "UserName"
+        elif "userPrincipalName" in user:
+            username_field = "userPrincipalName"
+        else:
+            return
 
-        access_control = await self._access_control_for_role_assignments(
-            site_web_url, list_item_role_assignments
-        )
+        prefixed_groups = set()
 
-        return self._decorate_with_access_control(list_item, access_control)
+        async for group in self.client.groups_user_transitive_member_of(
+            user[username_field]
+        ):
+            group_id = group["id"]
+            if group_id:
+                prefixed_groups.add(_prefix_group(group_id))
 
-    def _dls_enabled(self):
-        if self._features is None:
-            return False
+        email = user.get("EMail", user.get("mail", None))
+        username = user[username_field]
 
-        if not self._features.document_level_security_enabled():
-            return False
+        prefixed_mail = _prefix_email(email)
+        prefixed_username = _prefix_user(username)
+        id_ = email if email else username
 
-        return self.configuration["use_document_level_security"]
+        access_control = list({prefixed_mail, prefixed_username}.union(prefixed_groups))
 
-    def access_control_query(self, access_control):
-        return {"query": {"template": {"params": {"access_control": access_control}}}}
+        if "Modified" in user:
+            timestamp = datetime.strptime(user["Modified"], TIMESTAMP_FORMAT)
+        else:
+            timestamp = iso_utc()
 
-    def _user_access_control_doc(self, user, access_control):
         return {
-            "_id": user.get("Id"),
+            # Here we're intentionally using the email/username without the prefix
+            "_id": id_,
             "identity": {
-                "email": _prefix_email(user.get("Email")),
-                "username": _prefix_user(user.get("LoginName")),
+                "email": prefixed_mail,
+                "username": prefixed_username,
             },
+            "_timestamp": timestamp,
         } | self.access_control_query(access_control)
 
     async def get_access_control(self):
+        """Yields an access control document for every user of a site.
+        Note: this method can sometimes fetch some groups and therefore some users multiple times for different sites.
+        This can lead to a higher "indexed" number than actual documents present in the index (documents with the same id won't be indexed multiple times).
+        This is expected as it would otherwise overload memory keeping a set of all users/groups already seen for big Sharepoint Online instances.
```

Comment:
Probably would be worth to address this in UI at some point - I'm already constantly confused with number of docs in the index not being equal to number of docs ingested, etc. cc @danajuratoni

Reply:
Anyway addressed by caching emails/usernames, which were already seen -> reduces ingest time and number of calls to ES.